drivers/stepper/stepper_stm32.c

```diff
@@ -0,0 +1,757 @@
+	uint32_t max_steps;
+	/** Pending number of steps. */
+	uint32_t pending_steps;
+	/** PWM timer frequency (Hz). */
+	uint32_t f_tim;
+	/** Stepper motors data. */
+	struct stepper_stm32_motor_data *motors;
+};
+
+#define LL_TIM_CHANNEL_CH(n) UTIL_CAT(LL_TIM_CHANNEL_CH, n)
+#define LL_TIM_OC_SetCompareCH(n) UTIL_CAT(LL_TIM_OC_SetCompareCH, n)
+
+static inline struct stepper_stm32_data *to_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct stepper_stm32_config *to_config(struct device *dev)
+{
+	return dev->config_info;
+}
+
+static inline int32_t get_motor_index(const struct stepper_stm32_config *config,
+				      uint32_t channel)
+{
+	for (size_t i = 0u; i < config->n_motors; i++) {
+		if (config->motors[i].channel == channel) {
+			return (int32_t)i;
+		}
+	}
+
+	return -ENODEV;
+}
+
+/**
+ * @brief Obtain timer clock.
+ *
+ * @param clk	 Clock control device.
+ * @param pclken Timer clock details.
+ *
+ * @return Timer clock value (Hz).
+ */
+static uint32_t get_timer_clock(struct device *clk,
+				const struct stm32_pclken *pclken)
+{
+	int err;
+	uint32_t bus_clk, apb_psc, tim_clk;
+
+	err = clock_control_get_rate(clk, (clock_control_subsys_t *)pclken,
+				     &bus_clk);
+	__ASSERT_NO_MSG(err == 0);
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	if (pclken->bus == STM32_CLOCK_BUS_APB1) {
+		apb_psc = CONFIG_CLOCK_STM32_D2PPRE1;
+	} else {
+		apb_psc = CONFIG_CLOCK_STM32_D2PPRE2;
+	}
+
+	/*
+	 * Depending on pre-scaler selection (TIMPRE), timer clock frequency
+	 * is defined as follows:
+	 *
+	 * - TIMPRE=0: If the APB prescaler (PPRE1, PPRE2) is configured to a
+	 *   division factor of 1 then the timer clock equals to APB bus clock.
+	 *   Otherwise the timer clock is set to twice the frequency of APB bus
+	 *   clock.
+	 * - TIMPRE=1: If the APB prescaler (PPRE1, PPRE2) is configured to a
+	 *   division factor of 1, 2 or 4, then the timer clock equals to HCLK.
+	 *   Otherwise, the timer clock frequencies are set to four times to
+	 *   the frequency of the APB domain.
+	 */
+	if (LL_RCC_GetTIMPrescaler() == LL_RCC_TIM_PRESCALER_TWICE) {
+		if (apb_psc == 1U) {
+			tim_clk = bus_clk;
+		} else {
+			tim_clk = bus_clk * 2U;
+		}
+	} else {
+		if (apb_psc == 1U || apb_psc == 2U || apb_psc == 4U) {
+			tim_clk = SystemCoreClock;
+		} else {
+			tim_clk = bus_clk * 4U;
+		}
+	}
+#else
+	if (pclken->bus == STM32_CLOCK_BUS_APB1) {
+		apb_psc = CONFIG_CLOCK_STM32_APB1_PRESCALER;
+	}
+#if !defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_SOC_SERIES_STM32G0X)
+	else {
+		apb_psc = CONFIG_CLOCK_STM32_APB2_PRESCALER;
+	}
+#endif
+
+	/*
+	 * If the APB prescaler equals 1, the timer clock frequencies
+	 * are set to the same frequency as that of the APB domain.
+	 * Otherwise, they are set to twice (×2) the frequency of the
+	 * APB domain.
+	 */
+	if (apb_psc == 1U) {
+		tim_clk = bus_clk;
+	} else {
+		tim_clk = bus_clk * 2U;
+	}
+#endif
+
+	return tim_clk;
+}
```

Comment:
This function get_timer_clock() doesn't match all STM32 series cases.
For example some stm32f4 socs or stm32f7 socs have also x1 and x4 cases.
You can have a look at Arduino implementation which cover most cases:
https://github.com/stm32duino/Arduino_Core_STM32/blob/73f76d6028f7171f6273aa3a52cb0a4117f92ff8/cores/arduino/HardwareTimer.cpp#L1253

Reply:
Thanks, I'll take a look. PWM driver also needs a fix then, as this code is the same one.